### PR TITLE
Add message name is parameter and nonlocal

### LINKED ISSUE #6882: Add message `name-is-parameter-and-nonlocal`
### Current problem

The following produces a SyntaxError:
```python
def tomato(is_tasty: bool = True):
    nonlocal is_tasty
```

```
python example.py
  File "/Users/markbyrne/programming/example.py", line 2
    nonlocal is_tasty
    ^^^^^^^^^^^^^^^^^
SyntaxError: name 'is_tasty' is parameter and nonlocal
```

### Desired solution

Create a new message in Pylint and emit it in this scenario indicating the error.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ramonsaraiva**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Adds a new message for when a `nonlocal` name is also a parameter in a function scope

Closes #6882

### CODE REVIEW

**Comment by mbyrnepr2** [comment]
At first glance since this is a new check, could you also please add a good.py/bad.py example please? Here is where they live: https://github.com/PyCQA/pylint/tree/main/doc/data/messages

These can be tested locally with `tox test_doc`.

File: tests/functional/n/name/name_is_parameter_and_nonlocal.py
Comment (DanielNoord) [inline tests/functional/n/name/name_is_parameter_and_nonlocal.py:6]:
```suggestion
    nonlocal is_tasty # [name-is-parameter-and-nonlocal]


def tomato(*is_tasty):
    nonlocal is_tasty # [name-is-parameter-and-nonlocal]
```

I think this would also raise a syntax error, right?

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [inline pylint/checkers/base/basic_error_checker.py:210]:
```suggestion
            "Emitted when a nonlocal variable is also a parameter within a function defintion.",
```

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [inline pylint/checkers/base/basic_error_checker.py:208]:
```suggestion
            "Name %r is both parameter and nonlocal",
```

**Review by DanielNoord — CHANGES_REQUESTED** [review]
Could you also add some test with nested scopes and functions?

File: doc/user_guide/checkers/features.rst
Comment (Pierre-Sassoulas) [inline doc/user_guide/checkers/features.rst:68]: You're one of the very few contributors that actually generated the doc before opening a merge request, conratulation 👍  (Actually, you're probably the first one 😄 )

File: pylint/checkers/base/basic_error_checker.py
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/basic_error_checker.py:335]: We can probably optimize by using ``util.only_required_for_messages`` here.

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]

File: tests/functional/n/name/name_is_parameter_and_nonlocal.py
Comment (ramonsaraiva) [reply tests/functional/n/name/name_is_parameter_and_nonlocal.py:6]: I guess for that I can do something like
```python
        node_arg_names = {arg.name for arg in node.args.args}
        for local_name, local_nodes in node.locals.items():
            for local_node in local_nodes:
                if isinstance(local_node, nodes.Arguments):
                    node_arg_names.add(local_name)
```

Not sure if there's an easier way to retrieve locals that are considered arguments.

Comment (ramonsaraiva) [reply tests/functional/n/name/name_is_parameter_and_nonlocal.py:6]: For nested scopes and functions, are you looking for something like this?
```python
def grape(is_tasty: bool = True):
    nonlocal is_tasty # [name-is-parameter-and-nonlocal]

    def grape_juice(is_tasty: bool = True):
        nonlocal is_tasty  # [name-is-parameter-and-nonlocal]

    grape_juice()
```

Comment (DanielNoord) [reply tests/functional/n/name/name_is_parameter_and_nonlocal.py:6]: Doesn't `node.args.arguments()` work? According to the docstring that should be able to return all names that are used as parameters.

File: pylint/checkers/base/basic_error_checker.py
Comment (ramonsaraiva) [reply pylint/checkers/base/basic_error_checker.py:208]: I'm good to change to `.. is both parameter and nonlocal`, but python says `.. is parameter and nonlocal`:
```python
>>> def tomato(is_tasty):
...     nonlocal is_tasty
...
  File "<stdin>", line 2
SyntaxError: name 'is_tasty' is parameter and nonlocal
```
wdyt?

File: tests/functional/n/name/name_is_parameter_and_nonlocal.py
Comment (DanielNoord) [reply tests/functional/n/name/name_is_parameter_and_nonlocal.py:6]: Yeah, and variants of it. I haven't tested this out locally to see which combinations return a syntax error, but I think it is good to add some (regression tests) for possible combinations:
- outer scope with nonlocal inner scope with nonlocal
- outer scope without nonlocal inner scope with nonlocal
- outer scope with nonlocal inner scope without nonlocal
etc.

Scoping rules can create annoying false positives for messages like this.

File: pylint/checkers/base/basic_error_checker.py
Comment (DanielNoord) [reply pylint/checkers/base/basic_error_checker.py:208]: Hm, I'd say let's overwrite Python here and use more "readable" English. As pylint output is often read form IDEs I think we an be a little bit more verbose than the traditional error messages Python uses.

File: tests/functional/n/name/name_is_parameter_and_nonlocal.py
Comment (ramonsaraiva) [reply tests/functional/n/name/name_is_parameter_and_nonlocal.py:6]: > Doesn't `node.args.arguments()` work? According to the docstring that should be able to return all names that are used as parameters.

When I inspect `node.args` for a function that uses `(*is_tasty)`, it comes pretty much empty:
```python
(Pdb) vars(node.args)
{'lineno': None, 'col_offset': None, 'parent': <FunctionDef.watermelon l.9 at 0x7efe6e8078e0>, 'end_lineno': None, 'end_col_offset': None, 'position': None, 'vararg': 'is_tasty', 'kwarg': None, 'posonlyargs': [], 'posonlyargs_annotations': [], 'kwonlyargs_annotations': [], 'type_comment_args': [], 'type_comment_kwonlyargs': [], 'type_comment_posonlyargs': [], 'varargannotation': None, 'kwargannotation': None, 'args': [], 'defaults': [], 'kwonlyargs': [], 'kw_defaults': [], 'annotations': [], 'fromlineno': 9, 'tolineno': 9, 'arguments': []}
(Pdb) node.args.arguments
[]
```

File: pylint/checkers/base/basic_error_checker.py
Comment (ramonsaraiva) [reply pylint/checkers/base/basic_error_checker.py:208]: Totally agree.

File: tests/functional/n/name/name_is_parameter_and_nonlocal.py
Comment (ramonsaraiva) [reply tests/functional/n/name/name_is_parameter_and_nonlocal.py:6]: Sure! I can attempt that later today, this might also be related https://github.com/PyCQA/astroid/issues/761 - but not totally sure.

Comment (DanielNoord) [reply tests/functional/n/name/name_is_parameter_and_nonlocal.py:6]: Indeed let's not block this on too much refactoring and fix stuff as we run into them. I think this is a very clear use-case for a potential fix and I'd be happy to help release a patch version of `astroid` for it. I think there are other issues waiting on a patch as well, a regression in dataclasses for example.

**Comment by DanielNoord** [comment]
This seems like an easy PR for a maintainer to pick up.

@ramonsaraiva are you still planning to return to this? If not, would you mind if I picked it up?

**Comment by ramonsaraiva** [comment]
All good if you want to pick it up @DanielNoord - things got busier around here. I'll keep an eye on the solution!

**Comment by DanielNoord** [comment]
The tests here show a genuine failing false positive.

**Pull request closed without merging**
